prime-router/docs/design/validation/design.md

```diff
@@ -0,0 +1,284 @@
+![annotated-architecture-diagram.png](annotated-architecture-diagram.png)
+
+### Shared Components
+To facilitate reuse and extensibility, a new setting type `validationProfile` shall be created representing a type of
+validation, the validations to perform, and any applicable configuration parameters. This setting is reusable because
+senders and receivers can only be set up to receive one format at a time, so there should not be confusion over which
+validationProfile pertains to which format.
+
+#### Submission
+For FHIR, we currently use an `IParser` instance to parse and read some metadata. By default, the IParser performs some
+validation and raises warnings or errors (depending on how its configured). We need to investigate how the instance is
+currently configured and ensure it is performing validation. Ideally this configuration will be universal or binary
+(validation on/off vs different types) for all senders.
+
+This step will be a no-op for HL7.
+
+#### Pre transform
+Before we apply sender transforms and enrichment, we need to ensure that bundles contain sane data for its origin.
+We should also ensure any prerequisites (e.g. field, value, format) for transform/enrichment is satisfied. Further 
+validation is still necessary as the assembled bundle may still be missing data that will be added.
+
+#### Post transform
+Will occur during the convert function, after receiving and debatch, and before the convert step. We again need to 
+ensure that all necessary data exists and is sane. This will happen for both HL7 and FHIR data. One final validation will
+be needed (per receiver) after applying receiver transforms and enrichment.
+
+#### Pre dispatch
+Will occur in the translate function, after reception, translation, and enrichment but just before dispatch.
+
+Verify that the final dataset meets configured receiver expectations prior to dispatch.
+
+#### Actions Upon Validation Failure
+When a message fails validation, validation errors should show up in:
+- The submission history API
+- The Action Log in the database in such a way that the engagement team can easily query for it
+  - Submission validation failures should show up in Submission History UI
+    - Sender is responsible for issue triage escalating to ReportStream as necessary
+  - Send failures should show up in last-mile failures
+    - ReportStream is responsible for issue triage
+- Notify the sender immediately in API response if parse validation fails
+
+Tickets needed for both:
+- Create the validation profile setting: https://github.com/CDCgov/prime-reportstream/issues/9161
+- Add validation errors to the submission history API: https://github.com/CDCgov/prime-reportstream/issues/9036
+- Add validation errors to the action log: https://github.com/CDCgov/prime-reportstream/issues/9221
+
+### FHIR Validation
+A factory/builder shall consume the validationProfile setting and produce reusable, cached validation objects
```

Comment:
You don't say what tool(s)/libraries you're going to use for that here, should be explicit.

Reply:
@thetaurean Please address this.

Reply:
Was this addressed?